ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs

```diff
@@ -54,79 +54,79 @@ public void Dispose()
             if (!disposed)
             {
                 reader.Dispose();
-                stream.Dispose();
                 disposed = true;
             }
         }
 
         void ReadObjectCore()
         {
-            var type = ReadNextNodeType();
-            var name = ReadNullTerminatedString();
-            ReadValue(name, type);
+            KV1BinaryNodeType type;
+
+            // Keep reading values, until we reach the terminator
+            while ((type = ReadNextNodeType()) != KV1BinaryNodeType.End)
+                ReadValue(type);
         }
 
-        void ReadValue(string name, KV1BinaryNodeType type)
+        void ReadValue(KV1BinaryNodeType type)
         {
-            KVValue value;
-
-            switch (type)
+            // ASCII encoding is used for keys
+            var name = Encoding.ASCII.GetString(ReadNullTerminatedBytes());
+            if (type == KV1BinaryNodeType.ChildObject)
             {
-                case KV1BinaryNodeType.ChildObject:
-                    {
-                        listener.OnObjectStart(name);
-                        do
-                        {
-                            ReadObjectCore();
-                        }
-                        while (PeekNextNodeType() != KV1BinaryNodeType.End);
-                        listener.OnObjectEnd();
-                        return;
-                    }
-
-                case KV1BinaryNodeType.String:
-                    value = new KVObjectValue<string>(ReadNullTerminatedString(), KVValueType.String);
-                    break;
-
-                case KV1BinaryNodeType.WideString:
-                    throw new NotSupportedException("Wide String is not supported.");
-
-                case KV1BinaryNodeType.Int32:
-                case KV1BinaryNodeType.Color:
-                case KV1BinaryNodeType.Pointer:
-                    value = new KVObjectValue<int>(reader.ReadInt32(), KVValueType.Int32);
-                    break;
-
-                case KV1BinaryNodeType.UInt64:
-                    value = new KVObjectValue<ulong>(reader.ReadUInt64(), KVValueType.UInt64);
-                    break;
-
-                case KV1BinaryNodeType.Float32:
-                    var floatValue = BitConverter.ToSingle(reader.ReadBytes(4), 0);
-                    value = new KVObjectValue<float>(floatValue, KVValueType.FloatingPoint);
-                    break;
-
-                case KV1BinaryNodeType.Int64:
-                    value = new KVObjectValue<long>(reader.ReadInt64(), KVValueType.Int64);
-                    break;
-
-                default:
-                    throw new ArgumentOutOfRangeException(nameof(type));
+                listener.OnObjectStart(name);
+                ReadObjectCore();
+                listener.OnObjectEnd();
+            }
```

Comment:
If you `return` in the if, then you won't need to nest everything in the `else`.

Reply:
Don't see why this can't be done? You've struck it through.

Reply:
Implemented @yaakov-h's suggestion.